File: pkg/greenbrown/R/Phenology.R
Phenology <- structure(function(
	##title<< 
	## Calculate phenology metrics in time series
	##description<<
	## This function calculates from time series annual metrics of vegetation phenology:
	## \itemize{ 
	## \item{ \code{sos} start of season }
	## \item{ \code{eos} end of season }
	## \item{ \code{los} length of season }
	## \item{ \code{pop} position of peak value (maximum) }
	## \item{ \code{pot} position of trough value (minimum) }
	## \item{ \code{mgs} mean growing season value }
	## \item{ \code{peak} peak value (maximum) }
	## \item{ \code{trough} trough value (minimum) }
	## \item{ \code{msp} mean spring value }
	## \item{ \code{mau} mean autumn value }
	## \item{ \code{rsp} rate of spring greenup (not all methods) }
	## \item{ \code{rau} rate of autumn senescence rates (not all methods) }
	## }
	## The calculation of these metrics is performed in three steps and by using different methods:
	## \itemize{ 
	## \item{ Step 1: Filling of permanent (winter) gaps. See \code{\link{FillPermanentGaps}}}
	## \item{ Step 2: Time series smoothing and interpolation. See \code{\link{TsPP}} }	
	## \item{ Step 3: Detection of phenology metrics. Phenology metrics are estimated from the gap filled, smoothed and interpolated time series. This can be done by treshold methods (\code{\link{PhenoTrs}}) or by using the derivative of the time series (\code{\link{PhenoDeriv}}). }
	## }
	
	Yt, 
	### univariate time series of class \code{\link{ts}}
	
	approach = c("White", "Trs", "Deriv"), 
	### Approach to be used to calculate phenology metrics from smoothed time series. 'White' by sclaing annual cycles between 0 and 1 (White et al. 1997, see \code{\link{PhenoTrs}}); 'Trs' for simple tresholds (\code{\link{PhenoTrs}}); 'Deriv' by using the derivative of the smoothed function (\code{\link{PhenoDeriv}}).
	
	min.mean = 0.1,
	### minimum mean annual value in order to calculate phenology metrics. Use this threshold to suppress the calculation of metrics in grid cells with low average values	
	
	trs = NULL,
	### threshold to be used to determine SOS and EOS if method 'Trs' is used. If method 'Trs' is used but trs is NULL than trs will be computed from the long-term mean of Yt.
	
	fpg = FillPermanentGaps,
	### Filling of permanent gaps: If NULL, permanent gaps will be not filled, else the function \code{\link{FillPermanentGaps}} will be applied.
	
	tsgf = "TSGFspline",
	### Temporal smoothing and gap filling: Function to be used for temporal smoothing, gap filling and interpolation of the time series. If NULL, this step will be not applied. Otherwise a function needs to be specified. Exisiting functions that can be applied are \code{\link{TSGFspline}}, \code{\link{TSGFlinear}}, \code{\link{TSGFssa}}, \code{\link{TSGFdoublelog}}  
	
	interpolate = TRUE,
	### Should the smoothed and gap filled time series be interpolated to daily values?
	
	min.gapfrac = 0.2,
	### How often has an observation to be NA to be considered as a permanent gap? (fraction of time series length) Example: If the month January is 5 times NA in a 10 year time series (= 0.5), then the month January is considered as permanent gap if min.gapfrac = 0.4.	
	
	lower = TRUE,
	### For filling of permanent gaps: fill lower gaps (TRUE), upper gaps (FALSE) or lower and upper gaps (NULL)
	
	fillval = NA,
	### For filling of permanent gaps: constant fill values for gaps. If NA the fill value will be estimated from the data using fun. 
	
	fun = min,
	### For filling of permanent gaps: function to be used to compute fill values. By default, minimum.
	
	method = c("Elmore", "Beck"),
	### If 'tsgf' is TSGFdoublelog: Which kind of double logistic curve should be used to smooth the data? 'Elmore' (Elmore et al. 2012, see \code{\link{FitDoubleLogElmore}}) or 'Beck' (Beck et al. 2006, see \code{\link{FitDoubleLogBeck}}) .	
	
	backup = NULL,
	### Which backup algorithm should be used instead of TSGFdoublelog for temporal smoothing and gap filling if the time series has no seasonality? If a time series has no seasonal pattern, the fitting of double logistic functions is not meaningful. In this case another method can be used. Default: NULL (returns NA - no smoothing), other options: "TSGFspline", "TSGFssa", "TSGFlinear"	
	
	check.seasonality = 1:3,
	### Which methods in \code{\link{Seasonality}} should indicate TRUE (i.e. time series has seasonality) in order to calculate phenology metrics? 1:3 = all methods should indicate seasonality, Set to NULL in order to not perform seasonality checks.
	
	...
	### further arguments (currently not used)
	
	##details<<
	## This function allows to calculate phenology metrics on time series. This method  can be applied to gridded (raster) data using the function \code{\link{PhenologyRaster}}.
	
	##references<< 
	## Beck, P.S.A., C. Atzberger, K.A. Hodga, B. Johansen, A. Skidmore (2006): Improved monitoring of vegetation dynamics at very high latitudes: A new method using MODIS NDVI. - Remote Sensing of Environment 100:321-334. \cr
	## Elmore, A.J., S.M. Guinn, B.J. Minsley and A.D. Richardson (2012): Landscape controls on the timing of spring, autumn, and growing season length in mid-Atlantic forests. - Global Change Biology 18, 656-674. \cr	
	## White M.A., P.E. Thornton and S.W. Running (1997): A continental phenology model for monitoring vegetation responses to interannual climatic variability. - Global Biogeochemical Cycles 11, 217-234. 
	
	##seealso<<
	## \code{\link{PhenologyRaster}}, \code{\link{TSGFspline}}, \code{\link{TSGFssa}}, \code{\link{TSGFdoublelog}}, \code{\link{FitDoubleLogElmore}}, \code{\link{FitDoubleLogBeck}}

) {

	if (class(Yt) != "ts") stop("Yt should be from class 'ts'.")
	start <- start(Yt)
	freq <- frequency(Yt)
	nout <- 12 # number of output variables
	if (length(approach) > 1) approach <- approach[1]
	
	# time series pre-processing
	Yt1 <- TsPP(Yt, fpg=fpg, tsgf=tsgf, interpolate=interpolate, min.gapfrac=min.gapfrac, lower=lower, fillval=fillval, fun=fun, method=method, backup=backup)
	
	# check if the time series has Seasonality
	calc.pheno <- FALSE
	seasonal <- rep(FALSE, 3)
	if (!AllEqual(Yt1)) {
		if (is.null(check.seasonality)) {
			calc.pheno <- TRUE
		} else {
			seasonal <- Seasonality(Yt1)
			if (all(seasonal[check.seasonality])) calc.pheno <- TRUE # calculate phenology metrics only if 3 methods indicate seasonality
		}
	}
	
	# estimation of phenology metrics
	if (approach == "Trs") {	
		# treshold method based on mean treshold
		if (is.null(trs)) trs <- mean(Yt1, na.rm=TRUE)
		ts.agg <- aggregate(Yt1, FUN=PhenoTrs, approach="Trs", trs=trs, min.mean=min.mean, calc.pheno=calc.pheno)	
		
	} else if (approach == "White") {	
		# White et al. 1997
		ts.agg <- aggregate(Yt1, FUN=PhenoTrs, approach="White", trs=trs, min.mean=min.mean, calc.pheno=calc.pheno)	

	} else if (approach == "Deriv") {	
		# derivative method
		ts.agg <- aggregate(Yt1, FUN=PhenoDeriv, min.mean=min.mean, calc.pheno=calc.pheno)
	} else {
		stop(paste("Phenology: approach =", approach, "does not exist. Use 'Trs', 'White' or 'Deriv'."))
	}
		
	# reconstruct time series of metrics
	sos.ts <- ts(ts.agg[seq(1, length(ts.agg), by=nout)], start=start[1], frequency=1)
	eos.ts <- ts(ts.agg[seq(2, length(ts.agg), by=nout)], start=start[1], frequency=1)
	los.ts <- ts(ts.agg[seq(3, length(ts.agg), by=nout)], start=start[1], frequency=1)
	pop.ts <- ts(ts.agg[seq(4, length(ts.agg), by=nout)], start=start[1], frequency=1)
	pot.ts <- ts(ts.agg[seq(5, length(ts.agg), by=nout)], start=start[1], frequency=1)
	mgs.ts <- ts(ts.agg[seq(6, length(ts.agg), by=nout)], start=start[1], frequency=1)
	rsp.ts <- ts(ts.agg[seq(7, length(ts.agg), by=nout)], start=start[1], frequency=1)
	rau.ts <- ts(ts.agg[seq(8, length(ts.agg), by=nout)], start=start[1], frequency=1)
	peak.ts <- ts(ts.agg[seq(9, length(ts.agg), by=nout)], start=start[1], frequency=1)
	trough.ts <- ts(ts.agg[seq(10, length(ts.agg), by=nout)], start=start[1], frequency=1)
	msp.ts <- ts(ts.agg[seq(11, length(ts.agg), by=nout)], start=start[1], frequency=1)
	mau.ts <- ts(ts.agg[seq(12, length(ts.agg), by=nout)], start=start[1], frequency=1)
	
	# correct SOS, EOS and POP if they 'jump' over end of year and exclude outliers
	.correctDOY <- function(doy) {
		id <- 1:length(doy)
		atstart <- na.omit(id[doy <= 60]) # check if DOY is within first two monthy
		atend <- na.omit(id[doy >= 305]) # check if DOY is within last two months
		if (length(atstart) > 0 & length(atend) > 0) { # correct if some DOYs are at the end and some at the start of the year
			if (length(atstart) >= length(atend)) {
				doy[atend] <- doy[atend] - 365
			} else {
				doy[atstart] <- doy[atstart] + 365
			}
		}
		
		# exclude outliers if more than 60 days before or after median DOY
		med <- median(doy, na.rm=TRUE)
		rge <- IQR(doy, na.rm=TRUE) * 2
		doy[doy < (med - rge) | doy > (med + rge)] <- NA
		return(doy)
	}
	sos.ts <- .correctDOY(sos.ts)
	eos.ts <- .correctDOY(eos.ts)
	pop.ts <- .correctDOY(pop.ts)
	pot.ts <- .correctDOY(pot.ts)
	msp.ts[is.na(sos.ts)] <- NA
	mau.ts[is.na(eos.ts)] <- NA
	mgs.ts[is.na(sos.ts) & is.na(eos.ts)] <- NA
	los.ts[is.na(sos.ts) & is.na(eos.ts)] <- NA
	
	# prepare result
	phen <- list(
		series = Yt1,
		seasonal = seasonal,
		calc.pheno = calc.pheno,
		sos=sos.ts, 
		eos=eos.ts, 
		los=los.ts, 
		pop=pop.ts,
		pot=pot.ts,
		mgs=mgs.ts,
		peak=peak.ts,
		trough=trough.ts,
		msp=msp.ts,
		mau=mau.ts,
		rsp=rsp.ts,
		rau=rau.ts,
		freq.orig=freq,
		approach = approach,
		tsgf = tsgf
	)
	if (!is.null(tsgf)) {
		if (tsgf == "TSGFdoublelog") phen$method <- method[1]
	}
	class(phen) <- "Phenology"
	return(phen)
	### The function returns a "Phenology" object with the following components
	### \itemize{ 
	### \item{ \code{method} Selected method. }
	### \item{ \code{series} gap-filled, smoothed and daily interpolated time series }
	### \item{ \code{sos} annual time series of start of season }
	### \item{ \code{eos} annual time series of end of season }
	### \item{ \code{los} annual time series of length of season }
	### \item{ \code{pop} annual time series of position of peak (maximum) }
	### \item{ \code{pot} annual time series of position of trough (minimum) }
	### \item{ \code{mgs} annual time series of mean growing season values }
	### \item{ \code{peak} annual time series of peak value }
	### \item{ \code{trough} annual time series of trough value }
	### \item{ \code{msp} annual time series of mean spring value }
	### \item{ \code{mau} annual time series of mean autumn value }
	### \item{ \code{rsp} annual time series of spring greenup rates (only for methods 'Deriv' and 'Logistic')}
	### \item{ \code{rau} annual time series of autumn senescence rates (only for methods 'Deriv' and 'Logistic')}
	### }
}, ex=function() {
# load a time series of NDVI (normalized difference vegetation index)
data(ndvi)
plot(ndvi)

# introduce some missing values
winter <- (1:length(ndvi))[cycle(ndvi) == 1 | cycle(ndvi) == 2 | cycle(ndvi) == 12]
ndvi[sample(winter, length(winter)*0.5)] <- NA
plot(ndvi)

# spline fit and threshold
spl.trs <- Phenology(ndvi, tsgf="TSGFspline", approach="White") 
spl.trs
plot(spl.trs)	# default plot: start of season, end of season, position of peak
plot(spl.trs, type=c("los")) # length of season
plot(spl.trs, type=c("msp", "mgs", "mau", "peak")) # plot mean spring, growing season, autumn and peak values
plot(spl.trs$series, col="red"); lines(ndvi) # gap-filled and smoothed time series that was used to estimate phenology metrics

# calculate phenology metrics using different smoothing methods and approaches
lin.trs <- Phenology(ndvi, tsgf="TSGFlinear", approach="White") # linear interpolation/running median + treshold
lin.deriv <- Phenology(ndvi, tsgf="TSGFlinear", approach="Deriv") # linear interpolation/running median + derivativw
spl.trs <- Phenology(ndvi, tsgf="TSGFspline", approach="White") # spline + treshold
spl.deriv <- Phenology(ndvi, tsgf="TSGFspline", approach="Deriv") # spline + derivative
ssa.trs <- Phenology(ndvi, tsgf="TSGFssa", approach="White") # singular spectrum analysis + treshold
ssa.deriv <- Phenology(ndvi, tsgf="TSGFssa", approach="Deriv") # singular spectrum analysis + derivativw
beck.trs <- Phenology(ndvi, tsgf="TSGFdoublelog", method="Beck", approach="White") # double logistic fit + treshold
beck.deriv <- Phenology(ndvi, tsgf="TSGFdoublelog", method="Beck", approach="Deriv") # double logistic fit + derivative
elmore.trs <- Phenology(ndvi, tsgf="TSGFdoublelog", method="Elmore", approach="White") # double logistic fit + treshold
elmore.deriv <- Phenology(ndvi, tsgf="TSGFdoublelog", method="Elmore", approach="Deriv") # double logistic fit + derivative

# compare results: SOS and EOS
library(RColorBrewer)
type <- c("sos", "eos")
cols <- brewer.pal(10, "Paired")
nms <- c("Lin+Trs", "Lin+Deriv", "Spline+Trs", "Spline+Deriv", "SSA+Trs", "SSA+Deriv", "DoubleLog1+Trs", "DoubleLog1+Deriv", "DoubleLog2+Trs", "DoubleLog2+Deriv")
plot(lin.trs, col=cols[1], type=type, ylim=c(1, 365))
plot(lin.deriv, col=cols[2], type=type, add=TRUE)
plot(spl.trs, col=cols[3], type=type, add=TRUE)
plot(spl.deriv, col=cols[4], type=type, add=TRUE)
plot(ssa.trs, col=cols[5], type=type, add=TRUE)
plot(ssa.deriv, col=cols[6], type=type, add=TRUE)
plot(beck.trs, col=cols[7], type=type, add=TRUE)
plot(beck.deriv, col=cols[8], type=type, add=TRUE)
plot(elmore.trs, col=cols[9], type=type, add=TRUE)
plot(elmore.deriv, col=cols[10], type=type, add=TRUE)
legend("center", nms, text.col=cols, ncol=3, bty="n")

cor(cbind(lin.trs$sos, lin.deriv$sos, spl.trs$sos, spl.deriv$sos, ssa.trs$sos, ssa.deriv$sos, beck.trs$sos, beck.deriv$sos, elmore.trs$sos, elmore.deriv$sos), use="pairwise.complete.obs")
cor(cbind(lin.trs$eos, lin.deriv$eos, spl.trs$eos, spl.deriv$eos, ssa.trs$eos, ssa.deriv$eos, beck.trs$eos, beck.deriv$eos, elmore.trs$eos, elmore.deriv$eos), use="pairwise.complete.obs")

# compare results: LOS
type <- c("los")
plot(lin.trs, col=cols[1], type=type, ylim=c(130, 365))
plot(lin.deriv, col=cols[2], type=type, add=TRUE)
plot(spl.trs, col=cols[3], type=type, add=TRUE)
plot(spl.deriv, col=cols[4], type=type, add=TRUE)
plot(ssa.trs, col=cols[5], type=type, add=TRUE)
plot(ssa.deriv, col=cols[6], type=type, add=TRUE)
plot(beck.trs, col=cols[7], type=type, add=TRUE)
plot(beck.deriv, col=cols[8], type=type, add=TRUE)
plot(elmore.trs, col=cols[9], type=type, add=TRUE)
plot(elmore.deriv, col=cols[10], type=type, add=TRUE)
legend("bottom", nms, text.col=cols, ncol=5, bty="n")

# compare results: MSP, PEAK, TROUGH
type <- c("msp", "peak", "trough")
plot(lin.trs, col=cols[1], type=type, ylim=c(0.17, 0.37))
plot(lin.deriv, col=cols[2], type=type, add=TRUE)
plot(spl.trs, col=cols[3], type=type, add=TRUE)
plot(spl.deriv, col=cols[4], type=type, add=TRUE)
plot(ssa.trs, col=cols[5], type=type, add=TRUE)
plot(ssa.deriv, col=cols[6], type=type, add=TRUE)
plot(beck.trs, col=cols[7], type=type, add=TRUE)
plot(beck.deriv, col=cols[8], type=type, add=TRUE)
plot(elmore.trs, col=cols[9], type=type, add=TRUE)
plot(elmore.deriv, col=cols[10], type=type, add=TRUE)
legend("bottom", nms, text.col=cols, ncol=5, bty="n")


})

File: www/phenology.php

<!-- This is the project specific website template -->
<!-- It can be changed as liked or replaced by other content -->

<?php

$domain=ereg_replace('[^\.]*\.(.*)$','\1',$_SERVER['HTTP_HOST']);
$group_name=ereg_replace('([^\.]*)\..*$','\1',$_SERVER['HTTP_HOST']);
$themeroot='r-forge.r-project.org/themes/rforge/';

echo '<?xml version="1.0" encoding="UTF-8"?>';
?>
<!DOCTYPE html
	PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
	"http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en   ">

  <head>
	<meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
	<title><?php echo $group_name; ?></title>
	<link href="http://<?php echo $themeroot; ?>styles/estilo1.css" rel="stylesheet" type="text/css" />
	<link rel="stylesheet" href="styles.css" type="text/css">
  </head>

<body>

<!-- R-Forge Logo -->
<table border="0" width="100%" cellspacing="0" cellpadding="0">
<tr><td>
<a href="http://r-forge.r-project.org/"><img src="http://<?php echo $themeroot; ?>/imagesrf/logo.png" border="0" alt="R-Forge Logo" /> </a> </td> </tr>
</table>

<h1>greenbrown - land surface phenology and trend analysis</h1>

<!-- menu -->
<hr>
<h2>Contents</h2>
<ul>
	<li><a href="index.php">greenbrown: Introduction</a></li> 
	<li><a href="install.php">Installation and loading</a></li> 
	<li>Using greenbrown - applications and examples:</li>
	<ul>
		<li><a href="trends.php">Trends and breakpoints</a></li>
		<li><a href="phenology.php">Phenology analysis</a></li>
	</ul>
	<li><a href="http://<?php echo $domain; ?>/projects/<?php echo $group_name; ?>/">Developer summary page</a> </li>
</ul>
<hr>
<!-- end of menu -->

<h2>Using greenbrown: Phenology analysis</h2>

<p>Metrics of land surface phenology and greenness are often derived from vegetation index time series to map the spatial variability of vegetation phenology or to analyse temporal changes in vegetation. Phenology and greenness metrics (PGMs) are for example the start and end of growing season (SOS and EOS) or the greenness during the peak of the growing season (PEAK). A description of these metrics and a discussion on the caveats and challenges of phenology detection from remote sensing-derived vegetation index time series can be found in Forkel et al. (2015): </p>

<p><ul>Forkel M, Migliavacca M, Thonicke K, Reichstein M, Schaphoff S, Weber U, Carvalhais N (2015) <a href="http://onlinelibrary.wiley.com/doi/10.1111/gcb.12950/abstract">Co-dominant water control on global inter-annual variability and trends in land surface phenology and greenness.</a> Global Change Biology, accepted.</ul></p>

<p>In the greenbrown package, phenology detection is implemented as a step-wise procedure with prepocessing and analysis steps as described in Forkel et al. (2015). The steps are:</p>
<ol>
<li>Filling of permanent gaps in time series</li>
<li>Time series smoothing and interpolation</li>
<li>Detection of phenology events</li>
</ol>

<p>The corresponding functions to these steps can be used separately or the entire step-wise procedure can be applied just by using the main function <b><a href="man/Phenology.html">Phenology</a></b> for time series and <b><a href="man/PhenologyRaster.html">PhenologyRaster</a></b> for multi-temporal raster data. </p>

<p>The function <b><a href="man/Seasonality.html">Seasonality</a></b> is used during several steps to check if the time series has seasonality. Otherwise, the calculation of most phenology metrics like SOS and EOS is not useful.</p>

<h3>Main functions: Phenology and PhenologyRaster</h3>

<p>The main functions for phenology detection that cover these three steps are <b><a href="man/Phenology.html">Phenology</a></b> for time series and <b><a href="man/PhenologyRaster.html">PhenologyRaster</a></b> for multi-temporal raster data. These two functions encapsulate several other functions that are needed during the different steps and that are explained below.</p>

<blockquote>
> # load example NDVI time series<br>
> data(ndvi)<br>
> plot(ndvi)<br>
<br>
> # compute phenology metrics<br>
> spl.trs <- Phenology(ndvi, tsgf="TSGFspline", approach="White") <br>
> spl.trs<br>
--- Phenology ---------------------------- <br>
Calculate phenology metrics on time series <br>
------------------------------------------ <br>
<br>
Method <br>
 - smoothing and gap filling :  TSGFspline <br>
 - summary approach          :  White <br>
<br>
Mean and standard deviation of annual metrics: <br>
SOS    :  86 +- 11 <br>
EOS    :  280 +- 15 <br>
LOS    :  190 +- 21 <br>
MSP    :  0.27 +- 0.019 <br>
MAU    :  0.27 +- 0.019 <br>
RSP    :  NaN +- NA <br>
RAU    :  NaN +- NA <br>
POP    :  180 +- 10 <br>
POT    :  0.37 +- 1.3 <br>
MGS    :  0.3 +- 0.021 <br>
PEAK   :  0.32 +- 0.022 <br>
TROUGH :  0.23 +- 0.018<br>
<br>
> plot(spl.trs)
</blockquote>

<p><img src="fig_Phenology.png"><br>
Fig. 1: Annual time series of the the start of growing season (sos), end of growing season (eos) and position of peak day (pop).</p>

<p>The phenology analysis can be easily applied to raster datasets with <b><a href="man/PhenologyRaster.html">PhenologyRaster</a></b>.</p>

<blockquote>
> # load example NDVI time series<br>
> data(ndvimap)<br>
> phenmap <- PhenologyRaster(ndvimap, start=c(1982, 1), freq=12, tsgf="TSGFspline", approach="Deriv") <br>
> plot(phenmap, c(grep("SOS.1982", names(phenmap)), grep("EOS.1982", names(phenmap))))
</blockquote>

<p><img src="fig_PhenologyRaster.png"><br>
Fig. 2: Maps of start and end of the growing season in 1982 (in day of year).</p>


<h3>Step 1 - Filling of permanent gaps in time series: FillPermanentGap</h3>

<p>Permanent gaps in time series are missing values that occur regulary every year or in many years. Permanent gaps in satellite-derived vegetation index time series are for example the winter months in northern regions where reliable observation are not possible because of high sun zenit angles or snow cover. Such gaps can be identified using the function <b><a href="man/IsPermanentGap.html">IsPermanentGap</a></b> and these gaps can be filled with <b><a href="man/FillPermanentGaps.html">FillPermanentGaps</a></b>.</p>

<blockquote>
> data(ndvi)<br>
<br>
> # set NA values into winter months to simulate gaps<br>
> winter <- (1:length(ndvi))[cycle(ndvi) == 1 | cycle(ndvi) == 2 | cycle(ndvi) == 12]<br>
> gaps <- sample(winter, length(winter)*0.3)<br>
> ndvi2 <- ndvi<br>
> ndvi2[gaps] <- NA<br>
> plot(ndvi2)<br>
<br>
> # check and fill permanent gaps<br>
> IsPermanentGap(ndvi2)<br>
> fill <- FillPermanentGaps(ndvi2) # default fills with the minimum value<br>	
> plot(fill, col="red"); lines(ndvi)
</blockquote>


<h3>Step 2 - Temporal smoothing and gap filling: TsPP</h3>

<p>Several functions are implemented to smooth, to gap-fill and interpolate time series to daily time steps:<p>
<ul>
<li><b><a href="man/TSGFlinear.html">TSGFlinear</a></b>: linear interpolation and running median</li>
<li><b><a href="man/TSGFspline.html">TSGFspline</a></b>: spline smoothing and interpolation</li>
<li><b><a href="man/TSGFssa.html">TSGFssa</a></b>: singular spectrum analysis </li>
<li><b><a href="man/TSGFdoublelog.html">TSGFdoublelog</a></b>: fitting of a double-logistic function to the data of each year. Thereby two types of double-logistic function can be currently used: <b><a href="man/FitDoubleLogBeck.html">FitDoubleLogBeck</a></b> and <b><a href="man/FitDoubleLogElmore.html">FitDoubleLogElmore</a></b></li>
</ul>

<p>The functions of steps 1 and 2 are both encapsulated by the function <b><a href="man/TsPP.html">TsPP</a></b> (time series preprocessing).</p>

<blockquote>
> data(ndvi) <br>
<br>
> # introduce random gaps <br>
> gaps <- ndvi<br>
> gaps[runif(100, 1, length(ndvi))] <- NA<br>
<br>
> # smoothing and gap filling<br>
> tsgf <- TSGFspline(gaps)<br>
> plot(gaps, main="Temporal smoothing and gap-filling using splines")<br>
> lines(tsgf, col="red")<br>
</blockquote>

<p><img src="fig_TSGFspline.png"><br>
Fig. 3: Original time series with gaps (black) and smoothed and interpolated time series (red).</p>


<h3>Step 3 - Detection of phenology events: PhenoDeriv and PhenoTrs</h3>

<p>In the last step, phenology events like the start of growing season (SOS) and end of growing season (EOS) are estimated from the daily interpolated, gap-filled and smoothed time series. This can be done either by using thresholds (<b><a href="man/PhenoTrs.html">PhenoTrs</a></b>) or by extreme values of the derivative of the seasonal cycle (<b><a href="man/PhenoDeriv.html">PhenoDeriv</a></b>). <b><a href="man/PhenoTrs.html">PhenoTrs</a></b> implements thereby two approaches to estimate phenology events. <p> 

<blockquote>
> data(ndvi)<br>
<br>
> # time series pre-processing<br>
> x <- TsPP(ndvi, interpolate=TRUE)[1:365]<br>
<br>
> # calculate phenology metrics for first year<br>
> PhenoDeriv(x, plot=TRUE)
</blockquote>

<p><img src="fig_PhenoDeriv.png"><br>
Fig. 4: Seasonal cycle of a single year with estimated PGMs (phenology and greenness metrics).</p>


<br>
<hr>
<p><a href="index.php">greenbrown</a>, Matthias Forkel, mforkel [at] bgc-jena.mpg.de, 2015-04-15</p>

</body>
</html>
